// node-hill/dist/class/outfit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Outfit {
    constructor(player) {
        this._player = player;
        this._idString = new Set();
    }
    hat1(hatId) {
        this._player.assets.hat1 = hatId;
        this._idString.add("U");
        return this;
    }
    hat2(hatId) {
        this._player.assets.hat2 = hatId;
        this._idString.add("V");
        return this;
    }
    hat3(hatId) {
        this._player.assets.hat3 = hatId;
        this._idString.add("W");
        return this;
    }
    face(faceId) {
        this._player.assets.face = faceId;
        this._idString.add("Q");
        return this;
    }
    body(color) {
        this._player.colors.head = color;
        this._idString.add("K");
        this._player.colors.torso = color;
        this._idString.add("L");
        this._player.colors.rightArm = color;
        this._idString.add("N");
        this._player.colors.leftArm = color;
        this._idString.add("M");
        this._player.colors.leftLeg = color;
        this._idString.add("O");
        this._player.colors.rightLeg = color;
        this._idString.add("P");
        return this;
    }
    head(color) {
        this._player.colors.head = color;
        this._idString.add("K");
        return this;
    }
    torso(color) {
        this._player.colors.torso = color;
        this._idString.add("L");
        return this;
    }
    rightArm(color) {
        this._player.colors.rightArm = color;
        this._idString.add("N");
        return this;
    }
    leftArm(color) {
        this._player.colors.leftArm = color;
        this._idString.add("M");
        return this;
    }
    leftLeg(color) {
        this._player.colors.leftLeg = color;
        this._idString.add("O");
        return this;
    }
    rightLeg(color) {
        this._player.colors.rightLeg = color;
        this._idString.add("P");
        return this;
    }
    fromPlayer(player) {
        this._player.assets = Object.assign({}, player.assets);
        this._player.colors = Object.assign({}, player.colors);
        this._idString = new Set("UVWQKLNMOP");
        return this;
    }
    get idString() {
        return Array.from(this._idString).join("");
    }
    set() {
        return this._player.setOutfit(this);
    }
}
exports.default = Outfit;
